### PR TITLE
Fix url backslashes (invalid domains) like that in #3854 

### PR DESCRIPTION
**Pull request by OmarFarrag**
When trying to send a request, or even redirection, with url containing backslashes, it fails. You can see examples of that and how to reproduce the error in #3854. 

To fix that, I have added a downloader middleware called `URLMiddleware` that for now converts all backslashes in the url to forwardslashes, as this is the behavior of many browsers.

If confirmed, will immediately add tests and documentation

### CODE REVIEW

**Comment by elacuesta** [comment]
I humbly believe this should be addressed somewhere else, one possibility being [`w3lib.url.safe_url_string`](https://w3lib.readthedocs.io/en/latest/w3lib.html#w3lib.url.safe_url_string)

**Comment by OmarFarrag** [comment]
@elacuesta I thought that too, but w3lib won't support that behavior as it's not a standard (Some browsers do that, some don't) , so I think it should be handled by scrapy out of the box.. what do you think ?

**Comment by elacuesta** [comment]
`w3lib.url.safe_url_string` mentions RFC-3986, and the only reference to the backslash I could find in RFC-3986 was in section 7.3 (emphasis mine):

>Special care should be taken when the URI path interpretation process
involves the use of a back-end file system or related system
functions.  **File systems typically assign an operational meaning to
special characters, such as the "/", "\\", ":", "[", and "]"
characters**, and to special device names like ".", "..", "...", "aux",
"lpt", etc.  In some cases, merely testing for the existence of such
a name will cause the operating system to pause or invoke unrelated
system calls, leading to significant security concerns regarding
denial of service and unintended data transfer.  It would be
impossible for this specification to list all such significant
characters and device names.  Implementers should research the
reserved names and characters for the types of storage device that
may be attached to their applications and restrict the use of data
obtained from URI components accordingly.

Still, I'm not even sure I can interpret that as the backslash being an "unsafe character" :man_shrugging: 

That being said, there is already a PR to add additional functionality to `safe_url_string` which is not strictly related to RFC-3986 (https://github.com/scrapy/w3lib/pull/133).

TL;DR

I believe adding a middleware for this is a bit of overkill. Even if `safe_url_string` is not the right place to fix it, this could be done in the `Request` constructor, for instance. I know that different browsers seem to behave different here, but IMHO it doesn't make any sense to comply with a browser that chokes on an (admittedly invalid) URL.

**Comment by elacuesta** [comment]
Also, my understanding (I can't stress enough that this is my own personal opinion so I could be completely mistaken as to the philosophy of the library) is that `w3lib` provides helpers to deal with problems that the built-in `urllib` module doesn't solve, in _reasonable_ ways, even if that means sometimes go against some browsers. Like in this particular case, choking on an invalid URL is not reasonable to me if you can make it work by just replacing or removing a character (that shouldn't be there in the first place).
See also https://github.com/scrapy/scrapy/issues/3855#issuecomment-510559112

**Comment by OmarFarrag** [comment]
@elacuesta I believe you are right. I thought we couldn't modify w3lib code that's why I didn't think of going there, but now what you say seems more reasonable to me. 

Gonna close this PR..

**Comment by elacuesta** [comment]
@OmarFarrag I believe all projects under the Scrapy organization are open to contributions, so PRs are welcome. Thanks for taking an interest in this!
